tests/test_neoclassical.py

```diff
@@ -0,0 +1,179 @@
+"""Test for neoclassical transport compute functions."""
+
+from datetime import datetime
+
+import matplotlib.pyplot as plt
+import numpy as np
+import pytest
+from tests.test_plotting import tol_1d
+
+from desc.examples import get
+from desc.grid import LinearGrid
+from desc.utils import setdefault
+from desc.vmec import VMECIO
+
+
+@pytest.mark.unit
+def test_field_line_average():
+    """Test that field line average converges to surface average."""
+    rho = np.array([1])
+    alpha = np.array([0])
+    eq = get("DSHAPE")
+    iota_grid = LinearGrid(rho=rho, M=eq.M_grid, N=eq.N_grid, NFP=eq.NFP, sym=eq.sym)
+    iota = iota_grid.compress(eq.compute("iota", grid=iota_grid)["iota"]).item()
+    # For axisymmetric devices, one poloidal transit must be exact.
+    zeta = np.linspace(0, 2 * np.pi / iota, 25)
+    grid = eq.get_rtz_grid(rho, alpha, zeta, coordinates="raz")
+    data = eq.compute(["<L|r,a>", "<G|r,a>", "V_r(r)"], grid=grid)
+    np.testing.assert_allclose(
+        data["<L|r,a>"] / data["<G|r,a>"], data["V_r(r)"] / (4 * np.pi**2), rtol=1e-3
+    )
+    assert np.all(np.sign(data["<L|r,a>"]) > 0)
+    assert np.all(np.sign(data["<G|r,a>"]) > 0)
+
+    # Otherwise, many toroidal transits are necessary to sample surface.
+    eq = get("W7-X")
+    zeta = np.linspace(0, 40 * np.pi, 300)
+    grid = eq.get_rtz_grid(rho, alpha, zeta, coordinates="raz")
+    data = eq.compute(["<L|r,a>", "<G|r,a>", "V_r(r)"], grid=grid)
+    np.testing.assert_allclose(
+        data["<L|r,a>"] / data["<G|r,a>"], data["V_r(r)"] / (4 * np.pi**2), rtol=1e-3
+    )
+    assert np.all(np.sign(data["<L|r,a>"]) > 0)
+    assert np.all(np.sign(data["<G|r,a>"]) > 0)
+
+
+@pytest.mark.unit
+@pytest.mark.mpl_image_compare(remove_text=True, tolerance=tol_1d)
+def test_effective_ripple():
+    """Test effective ripple with W7-X."""
+    eq = get("W7-X")
+    rho = np.linspace(0, 1, 10)
+    alpha = np.array([0])
+    zeta = np.linspace(0, 20 * np.pi, 1000)
+    grid = eq.get_rtz_grid(rho, alpha, zeta, coordinates="raz")
+    data = eq.compute("effective ripple", grid=grid)
+    assert np.isfinite(data["effective ripple"]).all()
+    np.testing.assert_allclose(
+        data["epsilon 3/2"] ** (2 / 3),
+        data["effective ripple"],
+        err_msg="Bug in source grid logic in eq.compute.",
+    )
+    eps_32 = grid.compress(data["epsilon 3/2"])
+    fig, ax = plt.subplots()
+    ax.plot(rho, eps_32, marker="o")
+
+    neo_rho, neo_eps_32 = NeoIO.read("tests/inputs/neo_out.w7x")
+    np.testing.assert_allclose(
+        eps_32, np.interp(rho, neo_rho, neo_eps_32), rtol=0.16, atol=1e-5
```

Comment:
do we expect 16% error? Does increasing resolution help much?

Reply:
I've run some parameter scans and the difference persists; essentially like this

![image](https://github.com/user-attachments/assets/2bafd00e-c7c9-42ed-a840-9974f35f8f16)

1. I improved the bounce quadrature to have less clustering near edges after generating this plots, and at low resolution that has slightly increased  the match near rho =1/2 (can see by plotting the neo stuff on top in this test).
2. I haven't increased the Boozer resolution to generate the Neo plot. I've increased num_pitch for Neo, but my runs a few months ago would segfault on the cluster for some reason when I changed the other parameters for Neo

Reply:
One thing I don't like about the quadrature is that we don't reuse all the samples of the integrand we get from doing quadratures on the ripples interior to larger wells

Reply:
how well converged is DESC and NEO for this case? (does neo look the same if you *decrease* resolution?) Does desc chance if you increase any/all of the resolution parameters?

Reply:
There's minor variation but DESC converges to that plot above. If I decrease DESC resolution then things differ more. I've been working on things related to #1119 instead to shortcut the memory consumption of transforms rather than playing with neo or comparing to cases from Daniel's omnigenity paper.

Reply:
ok, I think we should look at this a bit more before merging. It could be due to the boozer resolution being too low? or idk if NEO artifiically truncates the boozer spectrum, I know a lot of codes ignore B_mn < 1e-5 or something, so that might also be giving small differences (and might explain why NEO looks smoother).
Would also be good to test against WISTELL, we'll see what they say or we can find them at APS

Reply:
that seems like a pretty significant improvement! is it just because bounce2d requires less memory so you can use higher resolution?

Reply:
also, is the vmec eq used for neo made from vmec, or exported from desc? if its originally from vmec that might explain the difference near the axis

Reply:
it was desc-solved W7X converted to wout

Reply:
> how well converged is DESC and NEO for this case? (does neo look the same if you decrease resolution?) Does desc chance if you increase any/all of the resolution parameters?

@f0uriest 
Again, when I decrease DESC resolution with the 1D methods, not really. With the 2D methods, when I decrease the resolution of only the FFTs that are used to compute the bounce integrals, then the results match Neo better. See https://github.com/PlasmaControl/DESC/pull/1290#discussion_r1811222246.

At sufficient resolution, both the 1D and 2D methods in DESC match.